### PR TITLE
fix: 8336 updated isLoading property in all widgets

### LINKED ISSUE #8336: [Feature] add an Animate Loading property for all widgets
## Summary
A property to enable or disable the loading animation for a particular widget whenever its data is getting updated. Today the platform does the animation automatically but a user should be able to opt out of it.

## Motivation
In scenarios where the animation does not make sense or if the data is getting polled, the user would want to disable this setting.

## Additional Context
Polling would get enabled with #8158

### PR DESCRIPTION
**Pull request by techbhavin**
## Description

> Add an Animate Loading property for all widgets

Fixes #8336

## Type of change

- New feature (non-breaking change which adds functionality)
## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented on my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:red_circle: Total coverage has decreased</summary>

    // Code coverage diff between base branch:release and head branch: feature/isLoading-property-all-widgets 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :red_circle: | total | 55.12 **(-0.01)** | 37.04 **(-0.01)** | 34.01 **(-0.01)** | 55.64 **(-0.01)**
 :red_circle: | app/client/src/sagas/WidgetLoadingSaga.ts | 25 **(-1.32)** | 15 **(-6.43)** | 14.29 **(-1.09)** | 29.82 **(-0.18)**
 :green_circle: | app/client/src/widgets/ModalWidget/widget/index.tsx | 38.89 **(0.69)** | 0 **(0)** | 16.67 **(0)** | 38.37 **(0.72)**</details>

### CODE REVIEW

**Review by jsartisan — APPROVED** [review]

**Comment by YogeshJayaseelan** [comment]
@techbhavin Verified this PR and please find the testing status below,

Issue 1 - Disabling the loading toggle for button widget and OnClick is binded to execute an API - Still the loading animation is displayed for button widget 

https://www.loom.com/share/c379da8ba7dd42e993182b6b67150f9f

Issue 2 - Loading animation does not happen for most of the widgets - ButtonGroup Widget, Icon Widget, Menu Button Widget, Image , Select , MultiSelect, MultiTreeSelect, TreeSelect, Switch, Radio, Ratting Widget 

https://www.loom.com/share/1e5b2d63d8864e32897c74ed16f4da62

**Review by jsartisan — DISMISSED** [review]

File: app/client/src/sagas/WidgetLoadingSaga.ts [hunk 1]
```diff
@@ -86,6 +93,15 @@ function* setWidgetsLoadingSaga() {
...
   );
 
+  // get all widgets evaluted data
+  const dataTree = yield select(getDataTree);
+  // check animateLoading is active on current widgets and set
+  Object.keys(entityDependencyMap).forEach((entity) => {
+    if (get(dataTree, [entity, "animateLoading"])) {
+      loadingEntities.add(entity);
+    }
+  });
```
Comment (hetunandu) [inline app/client/src/sagas/WidgetLoadingSaga.ts:103]: You need to check the `animateLoading` property and if it is false, remove it from the `loadingEntities` list.

**Review by hetunandu — CHANGES_REQUESTED** [review]

File: app/client/src/sagas/WidgetLoadingSaga.ts [hunk 2]
```diff
@@ -86,6 +94,18 @@ function* setWidgetsLoadingSaga() {
     new Set<string>(),
   );
 
+  // get all widgets evaluted data
+  const dataTree: DataTree = yield select(getDataTree);
+  // check animateLoading is active on current widgets and set
+  Object.entries(dataTree).forEach(([entityName, entity]) => {
+    if ("ENTITY_TYPE" in entity && entity.ENTITY_TYPE === ENTITY_TYPE.WIDGET)
+      if (get(dataTree, [entityName, "animateLoading"])) {
```
Comment (hetunandu) [inline app/client/src/sagas/WidgetLoadingSaga.ts:102]:
```suggestion
      if (get(dataTree, [entityName, "animateLoading"]) === false) {
```

File: app/client/src/sagas/WidgetLoadingSaga.ts [hunk 3]
```diff
@@ -86,6 +94,18 @@ function* setWidgetsLoadingSaga() {
...
+        loadingEntities.add(entityName);
+      } else {
```
Comment (hetunandu) [inline app/client/src/sagas/WidgetLoadingSaga.ts:104]:
```suggestion

```

**Review by hetunandu — CHANGES_REQUESTED** [review]

**Review by hetunandu — CHANGES_REQUESTED** [review]
By default, all animateLoading should be `true`

**Comment by YogeshJayaseelan** [comment]
@techbhavin Tested this PR and please find the testing status below,

Issue 1 - Not a valid scenario,
Issue 2 - Not a valid Scenario,

**New issue:**

Issue 3 - when the button group widget any of the button label name is binded with {{Query1.data.length}} and the same query is executed by clicking on the button widget - Button widget OnClick is binded to execute {{Query1.run()}} the animate loading is not happening for the button group widget.

https://www.loom.com/share/b4a599f520d44edcbbf57da057b29dfa

**Note : The above issue happens for Button group widget, Menu Widget**

Issue 4 - Animate loading toggle is not available for Switch group widget

Screenshot 2021-12-10 at 9 42 56 AM

**Comment by techbhavin** [comment]
@hetunandu  @YogeshJayaseelan  

issue 3: loading UI is not supported for Button group widget and Menu Widget
issue 4:   done

**Comment by YogeshJayaseelan** [comment]
@techbhavin Verified the fix and it is working as expected and we can proceed further

**Review by hetunandu — APPROVED** [review]

**Pull request merged**